jinahub/rankers/DPRReaderRanker/dpr_reader.py

```diff
@@ -0,0 +1,192 @@
+
+        self.default_traversal_paths = default_traversal_paths
+        self.default_batch_size = default_batch_size
+
+    @requests
+    def rank(
+        self, docs: Optional[DocumentArray], parameters: dict, **kwargs
+    ) -> DocumentArray:
+        """
+        Extracts answers from existing matches, (re)ranks them, and replaces the current
+        matches with extracted answers.
+
+        For each match ``num_spans_per_match`` of answers will be extracted, which
+        means that the new matches of the document will have a length of previous
+        number of matches times ``num_spans_per_match``.
+
+        The new matches will be have a score called ``relevance_score`` saved under
+        their scores. They will also have a tag ``span_score``, which refers to their
+        span score, which is used to rank answers that come from the same match.
+
+        If you specified ``title_tag_key`` at initialization, the tag ``title`` will
+        also be added to the new matches, and will equal the title of the match from
+        which they were extracted.
+
+        :param docs: Documents whose matches to re-rank (specifically, the matches of
+            the documents on the traversal paths will be re-ranked). The document's
+            ``text`` attribute is taken as the question, and the ``text`` attribute
+            of the matches as the context. If you specified ``title_tag_key`` at
+            initialization, the matches must also have a title (under this tag).
+        :param parameters: dictionary to define the ``traversal_path`` and the
+            ``batch_size``. For example::
+            parameters={'traversal_paths': ['r'], 'batch_size': 10}
+        """
+
+        if not docs:
+            return None
+
+        for doc in docs.traverse_flat(
+            parameters.get('traversal_paths', self.default_traversal_paths)
+        ):
+            if not doc.text:
+                self.logger.warning(
+                    f'No question (text) found for document with id {doc.id}; skipping'
+                    ' document. DPRReaderRanker requires a question from the main'
+                    ' document and context (text) from its matches.'
+                )
+                continue
+
+            new_matches = []
+
+            match_batches_generator = get_docs_batch_generator(
+                DocumentArray[doc],
+                traversal_path=['m'],
+                batch_size=parameters.get('batch_size', self.default_batch_size),
+                needs_attr='text',
+            )
+            for matches in match_batches_generator:
+                inputs = self._prepare_inputs(doc, matches)
+                with torch.no_grad():
+                    new_matches += self._get_new_matches(*inputs)
+
+            # Make sure answers are sorted by relevance scores
+            new_matches.sort(
+                key=lambda x: (x.scores['relevance_score'], x.scores['span_score']),
```

Comment:
I think it should be x.scores['relevance_score'].value

Reply:
You are right, these are some typos I didn't notice (will commit a fix in the next few mins)